suppressPackageStartupMessages(library("argparse"))
suppressPackageStartupMessages(library("data.table"))
suppressPackageStartupMessages(library("Seurat"))
suppressPackageStartupMessages(library("dplyr"))
suppressPackageStartupMessages(library("Matrix"))

# Helper function to dynamically identify and read gene/cell names

#' @param base_path The base path of the folder containing the files.
#' @param possible_names A vector of possible suffixes for gene or cell names files,
#' including common conventions like "genes", "features", "cells", and "barcodes".
#' @param extension A vector of possible file extensions, including both plain text 
#' and gzipped formats (".tsv", ".csv", ".tsv.gz", ".csv.gz").
find_path <- function(base_path, possible_names, extension){
  for (name in possible_names) {
    for (ext in extension) {
      full_path <- paste0(base_path, "/", name, ext)
      if (file.exists(full_path)) {
        # Automatically detect delimiter based on file extension
        delim <- ifelse(grepl(".csv", ext), ",", "\t")
        return(list(full_path, delim))
      }
    }
  }
  stop("Required gene or cell names file not found in: ", base_path)
}

#' @param sc_path Path to scRNA counts file (standard CytoSPACE input file format)
#' @param ct_path Path to cell type labels file (standard CytoSPACE input file format)
#' @param st_path Path to ST data (standard CytoSPACE input file format)
#' @param outdir Path to output directory
#' @param prefix Prefix of output file
#' @param disable_downsampling If TRUE, disables downsampling of scRNA-seq dataset for SCTransform() regardless of dataset size
#' @import data.table dplyr Seurat
get_cellfracs_seuratv3 <- function(sc_path, ct_path, st_path, outdir, prefix, disable_downsampling = FALSE){
    message(Sys.time(), " Load ST data")
    # Possible file names and extensions
    possible_gene_names <- c("genes", "features")
    possible_cell_names <- c("cells", "barcodes")
    possible_extensions <- c(".tsv", ".csv", ".tsv.gz", ".csv.gz")
    is_sparse_st <- endsWith(st_path, '.mtx') | endsWith(st_path, '.mtx.gz')
    if (is_sparse_st) {

        st <- Matrix::readMM(st_path)

        # cell and gene names expected to be stored and read from separate files
        base_dir <- dirname(st_path) 
        genes_path <- find_path(base_dir, possible_gene_names, possible_extensions)[[1]]
        cells_path <- find_path(base_dir, possible_cell_names, possible_extensions)[[1]]

        genes_delim <- find_path(base_dir, possible_gene_names, possible_extensions)[[2]]
        cells_delim <- find_path(base_dir, possible_cell_names, possible_extensions)[[2]]

        if(!file.exists(cells_path) | !file.exists(genes_path)) {
          stop("Please make sure to have the corresponding gene and cell names in separate files, 
                    present in the same directory as the matrix file, 
                    named as (genes/features)(.tsv/csv)(.gz) and (cells/barcodes)(.tsv/csv)(.gz)")
        }     

        cell_names <- read.csv(cells_path, sep = cells_delim, header = FALSE)
        gene_names <- read.csv(genes_path, sep = genes_delim, header = FALSE)

        if (!(dim(st)[1] == dim(gene_names)[1] & dim(st)[2] == dim(cell_names)[1])) {
            stop("The dimensions of the matrix do not match the number of genes and/or cells in the corresponding files.
                    The files for gene and cell lists must have one entry on each line, without a header line. Please check the files and try again.")
                    }
        
        # setting cell names as colnames and gene names as rownames of the matrix
        colnames(st) <- cell_names[[1]]
        rownames(st) <- gene_names[[1]]
        } else {
            if(endsWith(st_path,'.csv')){
                st_delim <- ','
            } else {
                st_delim <- '\t'
            }
            st <- fread(st_path, sep = st_delim, header = TRUE, data.table = FALSE)
            rownames(st) = st[,1]; st = st[,-1]
            st <- as.matrix(st)
    }

    
    st[is.na(st)] <- 0
    suppressMessages(st <- CreateSeuratObject(st) %>% SCTransform(verbose = FALSE, ncells = NULL) %>% RunPCA())
    
    message(Sys.time(), " Load scRNA data")
    is_sparse_sc <- endsWith(sc_path, '.mtx') | endsWith(sc_path, '.mtx.gz')
    if (is_sparse_sc) {

        scrna <- Matrix::readMM(sc_path)

        # cell and gene names expected to be stored and read from separate files
        base_dir <- dirname(sc_path) 
        genes_path <- find_path(base_dir, possible_gene_names, possible_extensions)[[1]]
        cells_path <- find_path(base_dir, possible_cell_names, possible_extensions)[[1]]

        genes_delim <- find_path(base_dir, possible_gene_names, possible_extensions)[[2]]
        cells_delim <- find_path(base_dir, possible_cell_names, possible_extensions)[[2]]

        if(!file.exists(cells_path) | !file.exists(genes_path)) {
          stop("Please make sure to have the corresponding gene and cell names in separate files, 
                    present in the same directory as the matrix file, 
                    named as (genes/features)(.tsv/csv)(.gz) and (cells/barcodes)(.tsv/csv)(.gz)")
        }     

        cell_names <- read.csv(cells_path, sep = cells_delim, header = FALSE)
        gene_names <- read.csv(genes_path, sep = genes_delim, header = FALSE)

        if (!(dim(scrna)[1] == dim(gene_names)[1] & dim(scrna)[2] == dim(cell_names)[1])) {
            stop("The dimensions of the matrix do not match the number of genes and/or cells in the corresponding files.
                    The files for gene and cell lists must have one entry on each line, without a header line. Please check the files and try again.")
                    }
        
        # setting cell names as colnames and gene names as rownames of the matrix
        colnames(scrna) <- cell_names[[1]]
        rownames(scrna) <- gene_names[[1]]
    } else {
        if(endsWith(sc_path,'.csv')){
            sc_delim <- ','
        } else {
            sc_delim <- '\t'
        }
        scrna <- fread(sc_path, sep = sc_delim, header = TRUE, data.table = FALSE)
        rownames(scrna) = scrna[,1]; scrna = scrna[,-1]
        scrna <- as.matrix(scrna)
    }
    
    scrna[is.na(scrna)] <- 0
    num_cells <- dim(scrna)[2]
    

    if(endsWith(ct_path,'.csv')){
        ct_delim <- ','
    } 
    else{
        ct_delim <- '\t'
    }

    celltypes <- fread(ct_path, sep = ct_delim, header = TRUE, data.table = FALSE)
    rownames(celltypes) = celltypes[,1]
    colnames(celltypes)[2] <- "CellType"
    celltypes <- celltypes[colnames(scrna),]

    scrna <- CreateSeuratObject(counts = scrna, meta.data = celltypes) 
    Idents(scrna) <- scrna$CellType
    cell_types <- unique(scrna$CellType)
    num_cell_types <- length(cell_types)
    names(cell_types) <- make.names(cell_types)
    Idents(scrna) <- make.names(Idents(scrna))
    if((num_cells > 10000) && (!disable_downsampling)) {
        warning("Downsampling to 10000 cells with equal number of cells for each cell type, for the purposes of cell type fraction estimation only. ",
                "This will not affect the dataset in the main CytoSPACE run. ",
                "To disable downsampling, please run get_cellfracs_seuratv3.R separately with the --disable-fraction-downsampling flag specified. ",
                "The output file [prefix]Seurat_weights.txt can then be passed to CytoSPACE with the -ctfep flag.")
        scrna <- subset(scrna, downsample = round(10000/num_cell_types))
    }
    if((dim(scrna)[2] < 300)) {
        warning("Please note that there may be an error in SCTransform() if there are too few cells available in the scRNA-seq dataset.")
    }
    suppressMessages(scrna <- SCTransform(scrna, verbose = FALSE, method = "glmGamPoi") %>% RunPCA())
    
    cell_index_vec <- Idents(scrna)
    message(Sys.time(), " Integration ")
    anchors <- FindTransferAnchors(reference = scrna, query = st, normalization.method = "SCT")
    predictions.assay <- TransferData(anchorset = anchors, refdata = cell_index_vec)
    colnames(predictions.assay) <- gsub('prediction.score.', '', colnames(predictions.assay))
    predictions.assay <- predictions.assay[, 2:(ncol(predictions.assay)-1), drop=FALSE]
    col_sums <- colSums(predictions.assay)
    cellfrac <- col_sums / sum(col_sums)
    cellfrac <- data.frame(Index = names(cellfrac), Fraction = cellfrac)
    cellfrac$Index <- cell_types[cellfrac$Index]

    write.table(predictions.assay, file.path(outdir, paste0(prefix, "Seurat_cellfracs.txt")), quote = FALSE, sep = "\t")
    write.table(t(cellfrac), file.path(outdir, paste0(prefix, "Seurat_weights.txt")), quote = FALSE, sep = "\t", col.names = FALSE)
} 



parser <- ArgumentParser()
# specify options 
parser$add_argument("--scrna-path", type="character", default="",
                    help="Path to single cell RNA-seq data, with rows as genes and columns as single cells")
parser$add_argument("--ct-path", type="character", default="",
                    help="Path to cell type annotation file, with first column showing single cell names, and second column showing cell type annotations")
parser$add_argument("--st-path", type="character", default="",
                    help="Path to Spatially-Resolved Transcriptomics (SRT) data, with rows as genes and columns as spots")
parser$add_argument("--outdir", type="character", default="./", help="Output directory")
parser$add_argument("--prefix", type="character", default="CytoSPACE_input.", help="Prefix for output files")
parser$add_argument("--disable-fraction-downsampling", action="store_true", help="Does not downsample scRNA-seq dataset if specified")


# get command line options, if help option encountered print help and exit,
# otherwise if options not found on command line then set defaults, 
args <- parser$parse_args()
get_cellfracs_seuratv3(sc_path = args$scrna_path, ct_path = args$ct_path, st_path = args$st_path,
                        outdir = args$outdir, prefix = args$prefix, disable_downsampling=args$disable_fraction_downsampling)
